crates/adapters/src/transport/durable/kafka/input.rs

```diff
@@ -0,0 +1,771 @@
+    #[serde(default)]
+    pub kafka_options: BTreeMap<String, String>,
+
+    /// Options passed to `rdkafka` for consumers only, as documented at
+    /// [`librdkafka`
+    /// options](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md).
+    ///
+    /// These options override `kafka_options` for consumers, and may be empty.
+    #[serde(default)]
+    pub consumer_options: BTreeMap<String, String>,
+
+    /// Options passed to `rdkafka` for producers only, as documented at
+    /// [`librdkafka`
+    /// options](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md).
+    ///
+    /// These options override `kafka_options` for consumers, and may be empty.
+    #[serde(default)]
+    pub producer_options: BTreeMap<String, String>,
+
+    /// Topic that contains the data to read.
+    pub data_topic: String,
+
+    /// Topic to use for recording the division of `data_topic` into steps.
+    /// Defaults to `<data_topic>_input-index`.
+    pub index_topic: Option<String>,
+
+    /// Partition to read.
+    pub partition: i32,
+
+    /// The log level of the client.
+    ///
+    /// If not specified, the log level will be calculated based on the global
+    /// log level of the `log` crate.
+    pub log_level: Option<KafkaLogLevel>,
+}
+
+#[derive(Clone)]
+struct Inner {
+    index_consumer_config: ClientConfig,
+    reader_consumer_config: ClientConfig,
+    producer_config: ClientConfig,
+    data_topic: String,
+    index_topic: String,
+    partition: i32,
+}
+
+impl Inner {
+    /// Set `option` to `val`; return an error if `option` is set to a different
+    /// value.
+    fn enforce_option<'a>(
+        settings: &mut BTreeMap<&'a str, &'a str>,
+        option: &'a str,
+        val: &'a str,
+    ) -> AnyResult<()> {
+        if *settings.entry(option).or_insert(val) != val {
+            bail!("cannot override '{option}' option: the Kafka transport adapter sets this option to '{val}'");
+        }
+        Ok(())
+    }
+
+    /// Set `option` to `val`, if missing.
+    fn set_option_if_missing<'a>(
+        settings: &mut BTreeMap<&'a str, &'a str>,
+        option: &'a str,
+        val: &'a str,
+    ) {
+        settings.entry(option).or_insert(val);
+    }
+
+    fn kafka_config(
+        source: &KafkaDurableInputConfig,
+        type_specific_options: &BTreeMap<String, String>,
+        overrides: &[(&str, &str)],
+        config_name: &str,
+    ) -> AnyResult<ClientConfig> {
+        let mut settings: BTreeMap<&str, &str> = source
+            .kafka_options
+            .iter()
+            .chain(type_specific_options.iter())
+            .map(|(o, v)| (o.as_str(), v.as_str()))
+            .collect();
+        for (option, val) in overrides {
+            Self::enforce_option(&mut settings, option, val)
+                .with_context(|| format!("Failed to validate Kafka options for {config_name}"))?;
+        }
+        let group_id = &Uuid::new_v4().to_string();
```

Comment:
This is super frustrating, because Kafka documents the group.id as only being needed if we want to use "consumer groups", which we don't, but in practice it just gives an error if you don't include it.

I don't have an answer to your question yet. It is a good question. I'll figure it out.

Reply:
I didn't realize this, I thought consumer groups were mandatory.

Reply:
I tested this and added a comment:

```patch
diff --git a/crates/adapters/src/transport/durable/kafka/input.rs b/crates/adapters/src/transport/durable/kafka/input.rs
index 19cda2af..9cceacca 100644
--- a/crates/adapters/src/transport/durable/kafka/input.rs
+++ b/crates/adapters/src/transport/durable/kafka/input.rs
@@ -182,9 +182,16 @@ impl Config {
             Self::enforce_option(&mut settings, option, val)
                 .with_context(|| format!("Failed to validate Kafka options for {config_name}"))?;
         }
+
+        // Set a unique `group.id` to ensure that we don't conflict with any
+        // existing consumer group.  In experiments, Kafka won't create any
+        // consumer group on the backend unless we implicitly (with
+        // `enable.auto.commit`) or explicitly commit an offset.  We don't do
+        // that, so this doesn't waste space on the Kafka brokers.
         let group_id = &Uuid::new_v4().to_string();
         Self::enforce_option(&mut settings, "group.id", group_id)
             .with_context(|| format!("Failed to validate Kafka options for {config_name}"))?;
+
         let default_redpanda_server = default_redpanda_server();
         Self::set_option_if_missing(&mut settings, "bootstrap.servers", &default_redpanda_server);
 
```